tests/pass_by_ref_test.py

```diff
@@ -447,3 +451,136 @@ def test_teal_output_is_unchanged(pt):
+
+    @Subroutine(TealType.uint64)
+    def not_ok(x: ScratchVar):
+        # not ok both at compile and runtime
+        return not_ok(x)
+
+    @Subroutine(TealType.uint64)
+    def not_ok_indirect1(x: ScratchVar):
+        return not_ok_indirect2(x)
+
+    @Subroutine(TealType.uint64)
+    def not_ok_indirect2(x: ScratchVar):
+        return not_ok_indirect1(x)
+
+    """
+    Complex subroutine graph example:
+
+    a --> b --> a (loop)
+            --> e 
+            --> f
+      --> *c--> g --> a (loop)
+            --> h 
+      --> d --> d (loop)
+            --> i
+            --> j
+
+    *c - this is the only "pass by-ref" subroutine
+    Expect the following error path: c-->g-->a-->c
+    """
+
+    @Subroutine(TealType.uint64)
+    def a(x):
+        tmp = ScratchVar(TealType.uint64)
+        return Seq(tmp.store(x), b(x) + c(tmp) + d(x))
+
+    @Subroutine(TealType.uint64)
+    def b(x):
+        return a(x) + e(x) * f(x)
+
+    @Subroutine(TealType.uint64)
+    def c(x: ScratchVar):
+        return g(Int(42)) - h(x.load())
+
+    @Subroutine(TealType.uint64)
+    def d(x):
+        return d(x) + i(Int(11)) * j(x)
+
+    @Subroutine(TealType.uint64)
+    def e(x):
+        return Int(42)
+
+    @Subroutine(TealType.uint64)
+    def f(x):
+        return Int(42)
+
+    @Subroutine(TealType.uint64)
+    def g(x):
+        return a(Int(17))
+
+    @Subroutine(TealType.uint64)
+    def h(x):
+        return Int(42)
+
+    @Subroutine(TealType.uint64)
+    def i(x):
+        return Int(42)
+
+    @Subroutine(TealType.uint64)
+    def j(x):
+        return Int(42)
+
+    ##### Approval PyTEAL Expressions #####
+
+    approval_ok = ok(Int(42))
+
+    x = ScratchVar(TealType.uint64)
+
+    approval_ok_byref = Seq(x.store(Int(42)), ok_byref(x))
+
+    approval_ok_indirect = ok_indirect1(Int(42))
+
+    approval_not_ok = Seq(x.store(Int(42)), not_ok(x))
+
+    approval_not_ok_indirect = Seq(x.store(Int(42)), not_ok_indirect1(x))
+
+    approval_its_complicated = a(Int(42))
+
+    ##### Assertions #####
+
+    assert compileTeal(approval_ok, Mode.Application, version=6)
+
+    assert compileTeal(approval_ok_byref, Mode.Application, version=6)
+
+    assert compileTeal(approval_ok_indirect, Mode.Application, version=6)
+
+    with pytest.raises(TealInputError) as err:
+        compileTeal(approval_not_ok, Mode.Application, version=6)
+
+    prefix = "ScratchVar arguments not allowed in recursive subroutines, but a recursive call-path was detected: "
+    assert f"{prefix}not_ok()-->not_ok()" in str(err)
+
+    with pytest.raises(TealInputError) as err:
+        compileTeal(approval_not_ok_indirect, Mode.Application, version=6)
+
+    assert (
+        f"{prefix}not_ok_indirect1()-->not_ok_indirect2()-->not_ok_indirect1()"
+        in str(err)
+    )
+
+    with pytest.raises(TealInputError) as err:
+        compileTeal(approval_its_complicated, Mode.Application, version=6)
+
+    assert f"{prefix}c()-->g()-->a()-->c()" in str(err)
```

Comment:
> I was not quite familiar with subroutines underlying implementation in `compiler` (until now). The implementation looks correct to me.
> 
> Just one question: it seems build is passing on `9e4bccd` but not `db0ea37`, while the latter just modified some lines in error messages. Do you know the reason for it?

The tests are failing because of the fact that I'm branching off `recursive-pass-by-ref-guardrails` which references a moving target on PySDK. Once this PR looks good to go code-wise, I will re-base against `master` and make sure the tests all pass.